### PR TITLE
As web authentication presentation context providing

### PR DESCRIPTION
**Pull request by james-rantmedia**
Since iOS 13, ASWebAuthenticationSession requires that a presentation context is provided in order to present the authentication UI. This context (`ASWebPresentationAnchor`) is typealiased to `UIWindow` on iOS.

This PR implements the necessary provider by returning the `authorizeContext` from `OAuath2Config`.

I also return the `error.asOAuth2Error` in the completion handler, as without it the web authentication session was failing silently on iOS 13.

### CODE REVIEW

**Comment by james-rantmedia** [comment]
The only new field I added was
```swift
/// Used to store the ASWebAuthenticationPresentationContextProvider
private var webAuthenticationPresentationContextProvider: AnyObject?
```
Which I added a doc line for the same as the other properties in that class. There were no changes necessary to the authConfig, as this uses the already existing `authorizeContext`.

This PR just exposes the `authorizeContext` to the new iOS 13 API which required a context to be passed.

**Comment by fotiDim** [comment]
I am getting some errors:
"generateCodeVerifier' is inaccessible due to 'internal' protection level"
"codeChallenge' is inaccessible due to 'internal' protection level"

It would be nice to get this fixed and merged soon

**Comment by james-rantmedia** [comment]
🤔

**Comment by fotiDim** [comment]
@james-rantmedia can you merge upstream master on your branch so that I can test?

**Comment by fotiDim** [comment]
@james-rantmedia actually I just forked the repo and tried to point SPM to my `master` and it is failing as well. Apparently unrelated to this PR but still interesting why it happens. 

Screenshot 2019-10-30 at 22 04 09

**Comment by fotiDim** [comment]
@james-rantmedia @larrybrunet I narrowed it down being the `36b154d5aae2d7ae121b8bdafbf86324ac5dc499` commit. The previous commit works just fine. Shall we open a new issue for this?

**Comment by ossus-lib** [comment]
Version 5.1.1 has these errors fixed!

**Comment by ernestoSk13** [comment]
Hi, is anybody checking this issue? seems like version 5.1.1 is not working, it failed in the CI

**Comment by ossus-lib** [comment]
@ernestoSk13 that issue had been fixe, it was just the Travis config that we needed to update.

**Comment by ykphuah** [comment]
I managed to get this branch to work, however, there's a fix needed for my code, before this, when I call authorizeEmbedded, I pass in the UIViewController as "from" parameter, which if I do useAuthenticationSession, it will complain that it is not a UIWindow.

So I need to pass in self.view.window as the "from" parameter, then it will work. I suggest to make this automatic in the pull request, so that passing either UIViewController or UIWindow will both work.

**Comment by gonzag** [comment]
I faced situations where the _authorizeContext_ is nil in **presentationAnchor**
Instead of crashing I've opted for return `ASPresentationAnchor()`:
```	
public func presentationAnchor(for session: ASWebAuthenticationSession) -> ASPresentationAnchor {
		guard let context = authorizer.oauth2.authConfig.authorizeContext as? ASPresentationAnchor else {
			return ASPresentationAnchor()
		}
		
		return context
	}
```

**Comment by ykphuah** [comment]
Actually it is not nil, its just not a ASPresentationAnchor which is UIWindow.

File: Sources/iOS/OAuth2Authorizer+iOS.swift [hunk 1]
```diff
@@ -142,9 +145,10 @@ open class OAuth2Authorizer: OAuth2AuthorizerUI {
 					self.oauth2.logger?.warn("OAuth2", msg: "Cannot intercept redirect URL: \(err)")
 				}
 			} else {
-				self.oauth2.didFail(with: nil)
+				self.oauth2.didFail(with: error?.asOAuth2Error)
```
Comment (reggian) [inline Sources/iOS/OAuth2Authorizer+iOS.swift:148]: I think that the `oauth2` depends on this to be `nil` in order to process and provide `OAuth2Error.requestCancelled` error.
Please see `Oauth2Base.didFail(with:)` method for details.

**Pull request merged**